crypto/compactcert/builder.go

```diff
@@ -0,0 +1,244 @@
+		parttree:     parttree,
+	}
+
+	return b, nil
+}
+
+// Present checks if the builder already contains a signature at a particular
+// offset.
+func (b *Builder) Present(pos uint64) bool {
+	return b.sigs[pos].part.Weight != 0
+}
+
+// Add a signature to the set of signatures available for building a certificate.
+// verifySig should be set to true in production; setting it to false is useful
+// for benchmarking to avoid the cost of signature checks.
+func (b *Builder) Add(pos uint64, sig crypto.OneTimeSignature, verifySig bool) error {
+	if b.Present(pos) {
+		return fmt.Errorf("position %d already added", pos)
+	}
+
+	// Check participants array
+	if pos >= uint64(len(b.participants)) {
+		return fmt.Errorf("pos %d >= len(participants) %d", pos, len(b.participants))
+	}
+
+	p := b.participants[pos]
+
+	if p.Weight == 0 {
+		return fmt.Errorf("position %d has zero weight", pos)
+	}
+
+	// Check signature
+	ephID := basics.OneTimeIDForRound(b.SigRound, p.KeyDilution)
+	if verifySig && !p.PK.Verify(ephID, b.Msg, sig) {
+		return fmt.Errorf("signature does not verify under ID %v", ephID)
+	}
+
+	// Remember the signature
+	b.sigs[pos].part = p
+	b.sigs[pos].Sig.OneTimeSignature = sig
+	b.signedWeight += p.Weight
+	b.cert = nil
+	return nil
+}
+
+// Ready returns whether the certificate is ready to be built.
+func (b *Builder) Ready() bool {
+	return b.signedWeight >= b.Params.ProvenWeight
+}
+
+// SignedWeight returns the total weight of signatures added so far.
+func (b *Builder) SignedWeight() uint64 {
+	return b.signedWeight
+}
+
+//msgp:ignore sigCommit
+type sigCommit []sigslot
+
+func (sc sigCommit) Length() uint64 {
+	return uint64(len(sc))
+}
+
+func (sc sigCommit) Get(pos uint64) (crypto.Hashable, error) {
+	if pos >= uint64(len(sc)) {
+		return nil, fmt.Errorf("pos %d past end %d", pos, len(sc))
+	}
+
+	return &sc[pos].sigslotCommit, nil
+}
+
+// coinIndex returns the position pos in the sigs array such that the sum
+// of all signature weights before pos is less than or equal to coinWeight,
+// but the sum of all signature weights up to and including pos exceeds
+// coinWeight.
+//
+// coinIndex works by doing a binary search on the sigs array.  The caller
+// should invoke it with lo=0 and hi=len(b.sigs).  The caller should make
+// sure that sigs[*].L is initialized before using coinIndex().
+func (b *Builder) coinIndex(coinWeight uint64, lo uint64, hi uint64) (uint64, error) {
+again:
+	if lo >= hi {
+		return 0, fmt.Errorf("coinIndex: lo %d >= hi %d", lo, hi)
+	}
+
+	mid := (lo + hi) / 2
+	if coinWeight < b.sigs[mid].L {
+		hi = mid
+		goto again
+	}
+
+	if coinWeight < b.sigs[mid].L+b.sigs[mid].part.Weight {
+		return mid, nil
+	}
+
+	lo = mid + 1
+	goto again
+}
+
+// Build returns a compact certificate, if the builder has accumulated
+// enough signatures to construct it.
+func (b *Builder) Build() (*Cert, error) {
+	if b.cert != nil {
+		return b.cert, nil
+	}
+
+	if b.signedWeight < b.Params.ProvenWeight {
```

Comment:
< should probably be <= (see https://github.com/algorand/go-algorand/pull/1359/files#r477474352)